### PR TITLE
fix: add this type for Reflect.defineProperty

### PR DESCRIPTION
**Pull request by hanzhangyu**
# Example
```ts
var A: {a?: number, b: number} = {b:2}
Reflect.defineProperty(A, 'a', {
  // set(this: any, value: any) { // to fix
  set(value: any) {
    console.log(this.b); // Property 'b' does not exist on type 'PropertyDescriptor'.
  },
  enumerable: true,
  configurable: true
});
A.a = 1;
```
# Fixes 
https://github.com/microsoft/TypeScript/issues/33542

### CODE REVIEW

**Comment by RyanCavanaugh** [comment]
This is not a good fix - just setting `this` to `any` is basically introducing an implicit `any` into the snippet.

**Comment by RyanCavanaugh** [comment]
Author pointed out that `this: any` is exactly what we do for `Object.defineProperty`, so at worse this is simply consistent with that

File: src/lib/es2015.reflect.d.ts [hunk 1]
```diff
@@ -1,7 +1,7 @@
 declare namespace Reflect {
     function apply(target: Function, thisArgument: any, argumentsList: ArrayLike<any>): any;
     function construct(target: Function, argumentsList: ArrayLike<any>, newTarget?: any): any;
-    function defineProperty(target: object, propertyKey: PropertyKey, attributes: PropertyDescriptor): boolean;
+    function defineProperty(target: object, propertyKey: PropertyKey, attributes: PropertyDescriptor & ThisType<any>): boolean;
```
Comment (Jessidhia) [inline src/lib/es2015.reflect.d.ts]: Why not

```ts
function defineProperty<T extends object>(
  target: T,
  propertyKey: PropertyKey,
  attributes: PropertyDescriptor & ThisType<T>
)
```

Comment (hanzhangyu) [reply src/lib/es2015.reflect.d.ts]: It doesn't make any sense, just try this case:
```ts
const obj = {a: 1};
defineProperty(obj, 'b', {
  get() {
    return this.b;
  }
})
```
The generics of typescript are completely static, they can\`t find the dynamic properties.

Comment (hanzhangyu) [reply src/lib/es2015.reflect.d.ts]: Should I mark this as a resolved conversation? @Jessidhia

**Review by sandersn — APPROVED** [review]

**Pull request merged**